[Salesforce] Editing for GA (#11254)
elastic/integrations: packages/salesforce/_dev/build/docs/README.md

## Overview

The Salesforce integration enables you to monitor your [Salesforce](https://www.salesforce.com/) instance. Salesforce is a customer relationship management (CRM) platform that supports businesses in managing marketing, sales, commerce, service, and IT teams from a unified platform accessible from anywhere.

You can use the Salesforce integration for:

> - **Operational insights**: Gain valuable insights into login and logout activities and other operational events within your organization.
> 
> - **Data visualization**: Create detailed visualizations to monitor, measure, and analyze usage trends and key data, helping you derive actionable business insights.
> 
> - **Proactive alerts**: Set up alerts to minimize Mean Time to Detection (MTTD) and Mean Time to Resolution (MTTR) by referencing relevant logs during troubleshooting.

Reviewer: ```suggestion
- **Operational insights**: Gain valuable insights into your organization's login and logout activities and other operational events.

- **Data visualization**: Create detailed visualizations to monitor, measure, and analyze usage trends and key data, helping you derive actionable business insights.

- **Proactive alerts**: Set up alerts to minimize Mean Time to Detection (MTTD) and Mean Time to Resolution (MTTR) by referencing relevant logs during troubleshooting.
```

## Data streams

The Salesforce integration collects the following data streams:

> - `login`: Collects information related to users who log in to Salesforce.
> - `logout`: Collects information related to users who log out from Salesforce.
> - `apex`: Collects information about various Apex events such as Callout, Execution, REST API, SOAP API, Trigger, and so on.
> - `setupaudittrail`: Collects information related to  changes users made in the organization's setup area for the last 180 days.

Reviewer: ```suggestion
- `login`: Collects information related to users who log in to Salesforce.
- `logout`: Collects information related to users who log out from Salesforce.
- `apex`: Collects information about various Apex events such as Callout, Execution, REST API, SOAP API, Trigger, and so on.
- `setupaudittrail`: Collects information related to changes users made in the organization's setup area for the last 180 days.
```

## Compatibility

This integration has been tested against the Salesforce Spring '22 (v54.0) release. The minimum supported version is v46.0.

To determine your Salesforce instance version, use one of the following methods:

- Salesforce Classic

  On the **Home** tab in Salesforce Classic, you can find a link in the top right corner that indicates the current release version of your Salesforce instance, for example `Summer '24 for Developers`.

- Use the Salesforce Instance URL

>   Use your Salesforce instance URL with the following format: `Salesforce Instance URL/services/data`, for example: https://na9.salesforce.com/services/data.

Reviewer: ```suggestion
  Use your Salesforce Instance URL with the following format: `<Salesforce Instance URL>/services/data`, for example: `https://na9.salesforce.com/services/data`, here `https://na9.salesforce.com` is the Salesforce Instance URL.
```

what do you think?

</Versions>
```

The last entry in the list indicates the current release version of your Salesforce instance. In this example, the version is `Summer '22 (v55.0)`.

## Prerequisites

> - You need Elasticsearch to store and search your data and Kibana to visualize and manage it. You can use our hosted Elasticsearch Service on Elastic Cloud, which is recommended or self-manage the Elastic Stack on your hardware.

Reviewer: ```suggestion
- You need Elasticsearch to store and search your data and Kibana to visualize and manage it. You can use our hosted Elasticsearch Service on Elastic Cloud, which is recommended, or self-manage the Elastic Stack on your hardware.
```

- Make sure API Enabled permission is selected for the user profile in your Salesforce instance:

  1. Go to **Setup** > **Quick Find** and type **Users**.
  2. Select **Users** from the left navigation tree.
  3. In the **Full Name** column, select the name associated with the user account used for data collection.
  4. Search for the **API Enabled** permission on the profile page. If it’s not present, search under **System Permissions** and check if the API Enabled privilege is selected. If not, enable it for data collection.

- Make sure that collecting data using [Real-Time Event Monitoring API](https://help.salesforce.com/s/articleView?id=sf.real_time_event_monitoring_enable.htm&type=5) is enabled:

>   1. Go to **Setup** > **Quick Find** and type **Event Manager**.
>   2. Select **Event Manager** from the left navigation tree.
>   3. To monitor an event, for example Login Event, or Logout Event, click the dropdown arrow and select **Enable Storage**.
>   4. Check if you have the required permissions: **View Real-Time Event Monitoring Data**.

Reviewer: ```suggestion
  1. Go to **Setup** > **Quick Find** and type **Event Manager**.
  2. Select **Event Manager** from the left navigation tree.
  3. To monitor an event, for example, Login Event, or Logout Event, click the dropdown arrow and select **Enable Storage**.
  4. Check if you have the required permissions: **View Real-Time Event Monitoring Data**.
```

### Salesforce instance URL

This is the URL of your Salesforce Organization.

- **Salesforce Classic**: Given the example URL https://na9.salesforce.com/home/home.jsp, the Salesforce Instance URL is extracted as https://na9.salesforce.com.

- **Salesforce Lightning**: The instance URL is available under your user name in the **View Profile** tab.

### Client key and client secret for authentication

To use this integration, you need to create a new Salesforce Application using OAuth. Follow these steps to create a connected application in Salesforce:

1. Log in to [Salesforce](https://login.salesforce.com/) with the user credentials you want to collect data with.
2. Click **Setup** in the top right menu bar.
3. In the **Search Setup** box, search for `App Manager` and select it.
4. Click **New Connected App**.
5. Provide a name for the connected application. This name will be displayed in the App Manager and on its App Launcher tile.
6. Enter the API name. The default is a version of the name without spaces. Only letters, numbers, and underscores are allowed. If the original app name contains any other characters, edit the default name.
7. Enter the contact email for Salesforce.

> 8. In the **API (Enable OAuth Settings)** section, select **Enable OAuth Settings**.
> 9. In the **Callback URL** field, enter the instance URL as described in [Salesforce instance URL](#salesforce-instance-url).

Reviewer: ```suggestion
8. Under the **API (Enable OAuth Settings)** section, check the box for **Enable OAuth Settings**. 
9. In the **Callback URL** field, enter the instance URL as specified in [Salesforce instance URL](#salesforce-instance-url).
```

> 14. Click **Continue**, then under **API details**, click **Manage Consumer Details**. Verify the user account using the Verification Code.
> 15. Copy the `Consumer Key` and `Consumer Secret` from the Consumer Details section. These should be populated as the values for Client ID and Client Secret, respectively, in the configuration.

Reviewer: ```suggestion
14. Click on **Continue**, then select **Manage Consumer Details** under **API details**. Verify the user account by entering the Verification Code.
15. Copy the `Consumer Key` and `Consumer Secret` from the Consumer Details section. These values should be used as the Client ID and Client Secret, respectively, in the configuration.
```

For more details, check the Salesforce documentation on how to [Create a Connected App](https://help.salesforce.com/apex/HTViewHelpDoc?id=connected_app_create.htm).

### Username

The User ID of the registered user.

### Password

The password used to authenticate the user.

### API version

To find the API version:

1. Go to `Setup` > `Quick Find` > `Apex Classes`.
2. Click `New`.
3. Click the `Version Settings` tab.
4. Refer to the `Version` dropdown for the API Version number.

## Validation

> Once the Salesforce integration is successfully configured, follows these steps to validate the setup:

Reviewer: ```suggestion
Once the Salesforce integration is successfully configured, follow these steps to validate the setup:
```

> 1. Navigate to the **Assets** tab within the Salesforce Integration. You should get a list of available dashboards corresponding to your configured data streams.
> 3. Select the dashboard relevant to your data stream (for example, login, logout, apex, setupaudittrail).
> 4. Verify that the dashboard is populated with the expected data.

Reviewer: ```suggestion
1. Navigate to the **Assets** tab in the Salesforce Integration. You will find a list of available dashboards related to your configured data streams.
2. Select the dashboard relevant to your data stream (for example, login, logout, apex, setupaudittrail).
3. Verify that the dashboard is populated with the expected data.
```

1. Data stream consolidation: Reduced from 6 to 4 data streams — `apex`, `login`, `logout`, and `setupaudittrail`.
2. Field mapping updates: Modified mappings for multiple fields.

> 3. Additional changes: Various other modifications to enhance the overall integration performance and functionality.

Reviewer: ```suggestion
3. Additional changes: Several other adjustments have been made to improve the overall performance and functionality of the integration.
```

### Data ingestion error

If you encounter data ingestion errors, you might get the following error message:

> oauth2 client: error loading credentials using user and password: oauth2: cannot fetch token: 400 Bad Request

**Solution:** Make sure that the `API Enabled` permission is granted to the `profile` associated with the `username` used for the integration. Check the [Prerequisites](#prerequisites) section for more information.

If the error persists, follow these steps:

> 1. Go to **Setup** > **Quick Find** > **App Manager**.
> 2. Find the app and click the corresponding arrow to check which actions are available.
> 3. Click **View**.
> 4. Get the key and secret by clicking **Manage Consumer Details** in the API section.
> 5. Click **Manage** to edit the policies.
> 6. Click **Edit Policies** and select **Relax IP restrictions** from the dropdown for IP Relaxation.

Reviewer: ```suggestion
1. Navigate to **Setup** > **Quick Find** > **App Manager**.
2. Locate the app and click the corresponding arrow to view available actions.
3. Click **View**.
4. Obtain the client key and secret by clicking on **Manage Consumer Details** in the API section.
5. Click on **Manage** to edit the policies.
6. Click on **Edit Policies** and choose **Relax IP restrictions** from the dropdown menu for IP Relaxation.
```
